clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -1800,11 +1888,181 @@ class SyclKernelDecompMarker : public SyclKernelFieldHandler {
         FD->addAttr(SYCLRequiresDecompositionAttr::CreateImplicit(
             SemaRef.getASTContext()));
       CollectionStack.back() = true;
+      PointerStack.pop_back();
+    } else if (PointerStack.pop_back_val()) {
+      // FIXME: Array of pointers/ array of type containing pointers
+      // will be handled in a follow up PR. Currently, they continue
+      // to trigger decomposition.
+      llvm_unreachable("PointerStack should not be true when handling arrays.");
     }
     return true;
   }
 };
 
+// This visitor is used to traverse a non-decomposed record/array to
+// generate a new type corresponding to this record/array.
+class SyclKernelPointerHandler : public SyclKernelFieldHandler {
+  llvm::SmallVector<CXXRecordDecl *, 8> ModifiedRecords;
+  SmallVector<CXXBaseSpecifier *, 8> ModifiedBases;
+
+  IdentifierInfo *getModifiedName(IdentifierInfo *Id) {
+    std::string Name =
+        Id ? (Twine("_generated_") + Id->getName()).str() : "_generated_";
```

Comment:
```suggestion
        Id ? (Twine("__generated_") + Id->getName()).str() : "__generated_";
```
I think two underscores are a bit safer.

Reply:
I used single underscore to remain consistent with other generated arguments/types/etc. For example we generate _arg_FieldName, _arg_base. I don't think the name really matters here actually. Because we only use it to copy to local clone. We even have arguments with same name sometimes

Reply:
Actually just remembered that __wrapper_class has 2 underscores. In any case though, I don't think it really matters because of reason I gave above. WDYT?

Reply:
I agree with you regarding `_arg_FieldName` and `_arg_base` since they are arguments of the generated kernel and therefore used only there.  But the new `_generated_*` type as well as `__wrapper_class` are types, so they are visible to the whole LLVM IR module. I would prefer all generated names to have two underscores since this naming is reserved for implementation and in case there is clash of names it is user's problem.

Reply:
Ok. I made the change